platform/default/online_file_source.cpp

```diff
@@ -176,6 +167,66 @@ class OnlineFileSource::Impl {
         }
     }
 
+    struct PendingRequests {
+        // initialize two pending queues: a high priority queue for "regular" requests and a low priority queue for offline download requests
+        // the high priority requests are in prioritizedPendingRequests*[0]
+        // the low priority requests are in prioritizedPendingRequests*[1]
+        // context: https://github.com/mapbox/mapbox-gl-native/issues/12655
+        PendingRequests() : prioritizedPendingRequestsList(2), prioritizedPendingRequestsMap(2) {}
+
+        std::vector<std::list<OnlineFileRequest*>> prioritizedPendingRequestsList;
+        std::vector<std::unordered_map<OnlineFileRequest*, std::list<OnlineFileRequest*>::iterator>> prioritizedPendingRequestsMap;
+
+        void remove(OnlineFileRequest* request) {
+            auto it = prioritizedPendingRequestsMap[0].find(request);
+            if (it != prioritizedPendingRequestsMap[0].end()) {
+                prioritizedPendingRequestsList[0].erase(it->second);
+                prioritizedPendingRequestsMap[0].erase(it);
+            }
+            else {
+                it = prioritizedPendingRequestsMap[1].find(request);
+                if (it != prioritizedPendingRequestsMap[1].end()) {
+                    prioritizedPendingRequestsList[1].erase(it->second);
+                    prioritizedPendingRequestsMap[1].erase(it);
```

Comment:
To reduce code duplication, you could wrap the list + map together in one class, and handle the prioritization outside of the class.